### PR TITLE
[Merged by Bors] - feat: port Data.FinEnum

### PR DESCRIPTION
**Pull request by zaxioms0**
---

Initial commit. All working, but need to clean up

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/FinEnum.lean [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+/-- create an exhaustive list of the values of a given type -/
+def toList (α) [FinEnum α] : List α :=
+  (List.finRange (card α)).map (Equiv).symm
+#align fin_enum.to_list FinEnum.toList
+
+open Function
+
+@[simp]
+theorem mem_to_list [FinEnum α] (x : α) : x ∈ toList α := by
```
Comment (jcommelin) [inline Mathlib/Data/FinEnum.lean:76]:
```suggestion
theorem mem_toList [FinEnum α] (x : α) : x ∈ toList α := by
```

File: Mathlib/Data/FinEnum.lean [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  simp [toList] ; exists Equiv x ; simp
+#align fin_enum.mem_to_list FinEnum.mem_to_list
+
+@[simp]
+theorem nodup_to_list [FinEnum α] : List.Nodup (toList α) := by
```
Comment (jcommelin) [inline Mathlib/Data/FinEnum.lean:81]:
```suggestion
theorem nodup_toList [FinEnum α] : List.Nodup (toList α) := by
```

File: Mathlib/Data/FinEnum.lean [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+#align fin_enum.psigma.fin_enum_prop_right FinEnum.Psigma.finEnumPropRight
+
+instance Psigma.finEnumPropProp {α : Prop} {β : α → Prop} [Decidable α] [∀ a, Decidable (β a)] :
+    FinEnum (Σ'a, β a) :=
+  if h : ∃ a, β a then ofList [⟨h.fst, h.snd⟩] (by rintro ⟨⟩ ; simp)
+  else ofList [] fun a => (h ⟨a.fst, a.snd⟩).elim
+#align fin_enum.psigma.fin_enum_prop_prop FinEnum.Psigma.finEnumPropProp
+
+instance (priority := 100) [FinEnum α] : Fintype α
+    where
```
Comment (jcommelin) [inline Mathlib/Data/FinEnum.lean:210]:
```suggestion
instance (priority := 100) [FinEnum α] : Fintype α where
```

File: Mathlib/Data/FinEnum.lean [hunk 4]
```diff
@@ -225,48 +226,47 @@ def Pi.tail {x : α} {xs : List α} (f : ∀ a, a ∈ (x :: xs : List α) → β
...
+    exists Pi.tail f
     constructor
     · apply xs_ih
     · ext (x h)
-      simp [pi.cons]
+      simp [Pi.cons]
       split_ifs
       subst x
       rfl
       rfl
```
Comment (jcommelin) [inline Mathlib/Data/FinEnum.lean:248]:
```suggestion
      · subst x
        rfl
      · rfl
```

**Review by jcommelin — COMMENTED** [review]
LGTM (2 minor comments)

bors d+

File: Mathlib/Data/FinEnum.lean [hunk 5]
```diff
@@ -38,58 +42,58 @@ namespace FinEnum
...
 def ofNodupList [DecidableEq α] (xs : List α) (h : ∀ x : α, x ∈ xs) (h' : List.Nodup xs) : FinEnum α
     where
   card := xs.length
   Equiv :=
-    ⟨fun x => ⟨xs.indexOf x, by rw [List.indexOf_lt_length] <;> apply h⟩, fun ⟨i, h⟩ =>
-      xs.nthLe _ h, fun x => by simp [of_nodup_list._match_1], fun ⟨i, h⟩ => by
-      simp [of_nodup_list._match_1, *] <;> rw [List.nthLe_index_of] <;> apply List.nodup_dedup⟩
+    ⟨fun x => ⟨xs.indexOf x, by rw [List.indexOf_lt_length] ; apply h⟩, fun ⟨i, h⟩ =>
+      xs.nthLe _ h, fun x => by simp [ofNodupList.match_1], fun ⟨i, h⟩ => by
+      simp [ofNodupList.match_1, *]⟩
```
[the remark this answers is not in the document]
Comment (zaxioms) [reply Mathlib/Data/FinEnum.lean:60]: no it doesn't :( 
It gives me
```
invalid 'simp', proposition expected
  [inst : DecidableEq α] → (xs : List α) → (∀ (x : α), x ∈ xs) → List.Nodup xs → FinEnum α
  ```
  
  Is this worth looking into more or should I just merge?

**Comment by zaxioms** [comment]
> v zaxioms can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

**Pull request closed without merging**